### ISSUE TITLE
System.Reflection.CustomAttributeFormatException: 'Value' property specified was not found

### ISSUE THREAD

**Issue by andronachev** [Azure/Azure-Functions#1463]
Since 1 day ago (somewhere in the last 12-24 hours) we started getting this weird exception our Prod environment: 
```

   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType, Boolean mustBeInheritable, IList derivedAttributes)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeFieldInfo field, RuntimeType caType)
   at System.Reflection.RuntimeFieldInfo.GetCustomAttributes(Boolean inherit)
   at ProtoBuf.Meta.AttributeMap.Create(MemberInfo member, Boolean inherit) in C:\projects\protobuf-net\src\protobuf-net\Meta\AttributeMap.cs:line 32
   at ProtoBuf.Meta.MetaType.NormalizeProtoMember(MemberInfo member, AttributeFamily family, Boolean forced, Boolean isEnum, List`1 partialMembers, Int32 dataMemberOffset, Boolean inferByTagName, EnumMember& enumMember, MemberInfo backingMember) in C:\projects\protobuf-net\src\protobuf-net\Meta\MetaType.cs:line 956
   at ProtoBuf.Meta.MetaType.ApplyDefaultBehaviour_AddMembers(AttributeFamily family, Boolean isEnum, List`1 partialMembers, Int32 dataMemberOffset, Boolean inferTagByName, ImplicitFields implicitMode, List`1 members, MemberInfo member, Boolean& forced, Boolean isPublic, Boolean isField, Type& effectiveType, List`1 enumMembers, MemberInfo backingMember) in C:\projects\protobuf-net\src\protobuf-net\Meta\MetaType.cs:line 791
   at ProtoBuf.Meta.MetaType.ApplyDefaultBehaviour() in C:\projects\protobuf-net\src\protobuf-net\Meta\MetaType.cs:line 721
   at ProtoBuf.Meta.RuntimeTypeModel.FindOrAddAuto(Type type, Boolean demand, Boolean addWithContractOnly, Boolean addEvenIfAutoDisabled) in C:\projects\protobuf-net\src\protobuf-net\Meta\RuntimeTypeModel.cs:line 616
   at ProtoBuf.Meta.RuntimeTypeModel.get_Item(Type type) in C:\projects\protobuf-net\src\protobuf-net\Meta\RuntimeTypeModel.cs:line 472
   at EventStore.ClientAPI.Messages.ClientMessage.InitializeSerializers()
--- End of stack trace from previous location where exception was thrown ---
   at EventStore.ClientAPI.EventStoreConnection.Create(ConnectionSettings connectionSettings, Uri uri, String connectionName)
   at EventStore.ClientAPI.EventStoreConnection.Create(String connectionString, ConnectionSettingsBuilder builder, String connectionName)
   at Plexe.Library.Infrastructure.EventStore.EventStoreOrgStore.GetAggregateEvents(Guid aggregateId, String aggregatTypeName)
   at Plexe.Library.Infrastructure.EventStore.EventStoreOrgStore.GetById[T](Guid aggregateId)
   at Plexe.Origination.Backend.Context.Events.ApplicationEventHandler.ChangeStatus(Func`2 test, Guid applicationId, String newStatus) in D:\a\1\s\src\Plexe.Origination.Backend.Context\Events\ApplicationEventHandler.cs:line 678
   at Plexe.Origination.Backend.Context.Events.ApplicationEventHandler.Handle(AccountingSystemConnectNotification event) in D:\a\1\s\src\Plexe.Origination.Backend.Context\Events\ApplicationEventHandler.cs:line 365
```

After doing some digging in both the EventStore.Org repositroy and the ProtoBuf repository and didn't see anything weird, I searched some more and encountered a couple of similar exceptions on github, and they were related to the Azure Function runtime version; 

https://github.com/Azure/Azure-Functions/issues/1201
https://github.com/Azure/app-service-announcements-discussions/issues/45
https://github.com/Azure/azure-functions-durable-extension/issues/913

.... looking a bit some more in the Runtime version, I notice as of right now (2020-01-22-09:40 UTC) we are running on PROD, the runtime version: **2.0.12961.0**   which i see on the Azure Function repo has been commited 7 days ago, and has a hotfix afterwards (but not sure if the hotfix has anything to do with it).

https://github.com/Azure/azure-functions-host/tags

Anyway, I now suspect this is some weird subtle DLL versioning conflict, and since other people had similar exceptions, I wanted to raise the flag: **could this be happening becuase of the azure function's runtime update ? **

I tried to specifically change the runtime version's configuration (FUNCTIONS_EXTENSION_VERSION ) to explicit version;  2.0.12888 , but it still happens.

Thanks for your feedback,
Vlad

**Comment by andronachev** [comment]
Hi,

I am getting back at this because I got no response - just thought to add more details with my findings so far. 

From the Exception stack trace we can see the problem revolves somehow around the **Protobuf-net** nuget package ( https://github.com/protobuf-net/protobuf-net ) that we are using in our Azure Function. We are using it indirectly, because we are actually using the Event Store nuget package ( https://github.com/EventStore/EventStore ) . 

However, we are using the **Event Store package version 5.0.1**  which in turn is depending on **Photobuf Package version 2.4.0**  . This is important because we also see in the exception stack trace, the following (I will actually paste the whole JSON exception and **bold** what is interesting) : 

{ 
   "ClassName":"System.Reflection.CustomAttributeFormatException",
   "Message":"'Value' property specified was not found.",
   "Data":null,
   "InnerException":{ 
      "ClassName":"System.Reflection.TargetInvocationException",
      "Message":"Exception has been thrown by the target of an invocation.",
      "Data":null,
      "InnerException":{ 
         "ClassName":"System.NotSupportedException",
         "Message":"Specified method is not supported.",
         "Data":null,
         "InnerException":null,
         "HelpURL":null,
         "StackTraceString":"   **at ProtoBuf.Internal.ThrowHelper.ThrowNotSupportedException() in C:\\projects\\protobuf-net\\src\\protobuf-net.Core\\Internal\\ThrowHelper.cs:line 58\r\n   at ProtoBuf.ProtoEnumAttribute.set_Value(Int32 value) in C:\\projects\\protobuf-net\\src\\protobuf-net.Core\\ProtoEnumAttribute.cs:line 23**",
         "RemoteStackTraceString":null,
         "RemoteStackIndex":0,
         "ExceptionMethod":null,
         "HResult":-2146233067,
         "Source":"**protobuf-net.Core**",
         "WatsonBuckets":""
      },
      "HelpURL":null,
      "StackTraceString":"   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)\r\n   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)\r\n   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType, Boolean mustBeInheritable, IList derivedAttributes)",
      "RemoteStackTraceString":null,
      "RemoteStackIndex":0,
      "ExceptionMethod":null,
      "HResult":-2146232828,
      "Source":"System.Private.CoreLib",
      "WatsonBuckets":""
   },
   "HelpURL":null,
   "StackTraceString":"   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType, Boolean mustBeInheritable, IList derivedAttributes)\r\n   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeFieldInfo field, RuntimeType caType)\r\n   at System.Reflection.RuntimeFieldInfo.GetCustomAttributes(Boolean inherit)\r\n   at ProtoBuf.Meta.AttributeMap.Create(MemberInfo member, Boolean inherit) in C:\\projects\\protobuf-net\\src\\protobuf-net\\Meta\\AttributeMap.cs:line 32\r\n   at ProtoBuf.Meta.MetaType.NormalizeProtoMember(MemberInfo member, AttributeFamily family, Boolean forced, Boolean isEnum, List`1 partialMembers, Int32 dataMemberOffset, Boolean inferByTagName, EnumMember& enumMember, MemberInfo backingMember) in C:\\projects\\protobuf-net\\src\\protobuf-net\\Meta\\MetaType.cs:line 956\r\n   at ProtoBuf.Meta.MetaType.ApplyDefaultBehaviour_AddMembers(AttributeFamily family, Boolean isEnum, List`1 partialMembers, Int32 dataMemberOffset, Boolean inferTagByName, ImplicitFields implicitMode, List`1 members, MemberInfo member, Boolean& forced, Boolean isPublic, Boolean isField, Type& effectiveType, List`1 enumMembers, MemberInfo backingMember) in C:\\projects\\protobuf-net\\src\\protobuf-net\\Meta\\MetaType.cs:line 791\r\n   at ProtoBuf.Meta.MetaType.ApplyDefaultBehaviour() in C:\\projects\\protobuf-net\\src\\protobuf-net\\Meta\\MetaType.cs:line 721\r\n   at ProtoBuf.Meta.RuntimeTypeModel.FindOrAddAuto(Type type, Boolean demand, Boolean addWithContractOnly, Boolean addEvenIfAutoDisabled) in C:\\projects\\protobuf-net\\src\\protobuf-net\\Meta\\RuntimeTypeModel.cs:line 616\r\n   at ProtoBuf.Meta.RuntimeTypeModel.get_Item(Type type) in C:\\projects\\protobuf-net\\src\\protobuf-net\\Meta\\RuntimeTypeModel.cs:line 472\r\n   at EventStore.ClientAPI.Messages.ClientMessage.InitializeSerializers()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at EventStore.ClientAPI.EventStoreConnection.Create(ConnectionSettings connectionSettings, Uri uri, String connectionName)\r\n   at EventStore.ClientAPI.EventStoreConnection.Create(String connectionString, ConnectionSettingsBuilder builder, String connectionName)\r\n   at Plexe.Library.Infrastructure.EventStore.EventStoreOrgStore.GetAggregateEvents(Guid aggregateId, String aggregatTypeName)\r\n   at Plexe.Library.Infrastructure.EventStore.EventStoreOrgStore.GetById[T](Guid aggregateId)\r\n   at Plexe.Origination.Backend.QueryFunction.OriginationTrigger.PingEventStore(HttpRequest req, ILogger log2) in d:\\a\\1\\s\\src\\Plexe.Origination.Backend.QueryFunction\\OriginationTrigger.cs:line 42",
   "RemoteStackTraceString":null,
   "RemoteStackIndex":0,
   "ExceptionMethod":null,
   "HResult":-2146232827,
   "Source":"System.Private.CoreLib",
   "WatsonBuckets":"....."
}

It is important to notice we have stacktrace from the **protobuf-net.Core** namespace which **doesn't exist in the photobuf-net version 2.4.0** . I have also downloaded our azure function's content from the azure portal just to make sure, decompiled the photobuf-net DLL that I found in the bin folder and it was the right version, version 2.4.0 , that has no CORE namspace and no C# code in the ProtoEnumAttribute.cs file at line 23 that could throw this exception. 

Upon looking in the photobuf repository, **surprise**, it contains the protobuf-net.Core namespace as well as, if we look in the ProtoEnumAttribute.cs class, line 23, there is the bit of C# code that gives this exception: 

https://github.com/protobuf-net/protobuf-net/blob/master/src/protobuf-net.Core/ProtoEnumAttribute.cs

But we don't use Photobuf latest version, we use version 2.4.0, double checked our Azure Function bin content, event decompiled because i got that paranoid, so we are for sure not having any trace of this Photobuf latest addition in our bin folder. **Where does it come from ? **

As I said in the initial post, I am at the moment suspecting this is due to the Azure Function's Host being updated to latest runtime version: 2.0.12961.0  . Therefore I got this version's source code from the Azure repository and just searched for the protobuf-net dependency. Surprise again, lots of dependencies on **protobuf-net version 3.0.0-alpha.91**  **which contains the Protobuf-net.Core namespace** and also the problematic code. Screenshot below: 

![image](https://user-images.githubusercontent.com/16139612/73210244-83a56600-4152-11ea-96c9-b6cefb9b9804.png)

Of course I am not familiar with the Function's Host source code but I can only assume this has something to do with the exception we are getting. 

Also in addition to this overall I want to mention that we are using the same C# code that is throwing this exception in other contexts which are not hosted in an Azure Function (they are Web app service), and in the Web App Service we are not having any issues with it. Nonetheless this is an urgent matter for us.

**TL;DR:** could be that the latest Function's Host dependency on the 3.0.0-alpha91 version of protobuf-net is somehow interfering with our own dependency on Protobuf-net version 2.4.0 and we are getting this exception which is possible to happen in version 3.0.0-alpha91 but **not in 2.4.0**


Thoughts ? 
Vlad

**Comment by v-bbalaiagar** [comment]
Closing this issue as the fix has been globally deployed [here](https://github.com/Azure/azure-functions-host/issues/5414). Please feel free to reach back to us in case of any further queries related to this issue.
